When using Discovery SPI, cluster join only happens via split-brain mechanism

<!--
Thanks for reporting your issue. Please share with us the following information, to help us resolve your issue quickly and efficiently.
-->
  
**Describe the bug**
When using `TcpIpJoiner` or `MulticastJoin` methods, everything works correctly,
but when using `networkConfig.getJoin().getDiscoveryConfig().setDiscoveryServiceProvider()` mechanism,
the cluster isn't joined, and the freshly-started node only sees itself.
However, when split brain timeout is reached, the split-brain mechanism is used to join the cluster
NOTE: This only happens when `WAIT_SECONDS_BEFORE_JOIN` is set to zero.

**Expected behavior**
Cluster join should happen on startup time instead of waiting for split-brain logic

**To Reproduce**
- pull https://github.com/flowlogix/hazelcast-issues
- in the directory CacheTester run `mvn -Prun test`
- do the same in an additional window and watch the logs

**Additional context**
Hazelcast version is 4.1-SNAPSHOT (latest)
Java 8
Here are the logs from the second node. First node works as expected:

```
Sep 20, 2020 9:01:18 PM com.hazelcast.instance.AddressPicker
INFO: [LOCAL] [dev] [4.1-SNAPSHOT] Using public address: [127.0.0.1]:5711
Sep 20, 2020 9:01:18 PM com.hazelcast.system
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Hazelcast 4.1-SNAPSHOT (20200920 - 6ed5868) starting at [127.0.0.1]:5711
Sep 20, 2020 9:01:18 PM com.hazelcast.instance.impl.Node
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Using Discovery SPI
Sep 20, 2020 9:01:18 PM com.hazelcast.cp.CPSubsystem
WARNING: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
Sep 20, 2020 9:01:19 PM com.hazelcast.internal.diagnostics.Diagnostics
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
Sep 20, 2020 9:01:19 PM com.hazelcast.core.LifecycleService
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5711 is STARTING
Sep 20, 2020 9:01:19 PM com.hazelcast.internal.cluster.ClusterService
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] 

Members {size:1, ver:1} [
	Member [127.0.0.1]:5711 - 78f83873-16c2-47e4-8d28-1993c778e9c4 this
]

Sep 20, 2020 9:01:19 PM com.hazelcast.instance.impl.Node
WARNING: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Config seed port is 5710 and cluster size is 1. Some of the ports seem occupied!
Sep 20, 2020 9:01:19 PM com.hazelcast.core.LifecycleService
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5711 is STARTED
<c> to create cache, <d> to destroy, <p> to print, <l> to lock, <u> to unlock
blank to exit <h> for help, anything else to put in cache ...

*** initial startup, the node sees only itself ***

Sep 20, 2020 9:01:24 PM com.hazelcast.internal.server.tcp.TcpServerConnection
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Initialized new cluster connection between /127.0.0.1:64664 and /127.0.0.1:5710
Sep 20, 2020 9:01:24 PM com.hazelcast.internal.cluster.impl.ClusterJoinManager
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] We should merge to [127.0.0.1]:5710, both have the same data member count: 1

*** now split-brain merge mechanism activated after the merge timeout ***

Sep 20, 2020 9:01:24 PM com.hazelcast.internal.cluster.impl.DiscoveryJoiner
WARNING: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5711 is merging [tcp/ip] to [127.0.0.1]:5710
Sep 20, 2020 9:01:24 PM com.hazelcast.internal.cluster.impl.operations.LockClusterStateOp
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Locking cluster state. Initiator: [127.0.0.1]:5711, lease-time: 60000
Sep 20, 2020 9:01:24 PM com.hazelcast.internal.cluster.impl.operations.LockClusterStateOp
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Extending cluster state lock. Initiator: [127.0.0.1]:5711, lease-time: 20000
Sep 20, 2020 9:01:24 PM com.hazelcast.internal.cluster.impl.operations.CommitClusterStateOp
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Changing cluster state from ClusterState{state=ACTIVE, lock=LockGuard{lockOwner=[127.0.0.1]:5711, lockOwnerId='7972e4d0-186e-45ca-bbf2-3f2ae64e4173', lockExpiryTime=1600653764043}} to ClusterStateChange{type=class com.hazelcast.cluster.ClusterState, newState=FROZEN}, initiator: [127.0.0.1]:5711, transient: false
Sep 20, 2020 9:01:24 PM com.hazelcast.internal.cluster.impl.operations.MergeClustersOp
WARNING: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5711 is merging to [127.0.0.1]:5710, because: instructed by master [127.0.0.1]:5711
Sep 20, 2020 9:01:24 PM com.hazelcast.core.LifecycleService
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5711 is MERGING
Sep 20, 2020 9:01:24 PM com.hazelcast.internal.cluster.ClusterService
WARNING: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Resetting local member UUID. Previous: 78f83873-16c2-47e4-8d28-1993c778e9c4, new: fbe0e9ae-29f9-4d04-9c7b-231fbd569316
Sep 20, 2020 9:01:24 PM com.hazelcast.internal.server.tcp.TcpServerConnection
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Connection[id=1, /127.0.0.1:64664->/127.0.0.1:5710, qualifier=null, endpoint=[127.0.0.1]:5710, alive=false, connectionType=MEMBER] closed. Reason: TcpServer is stopping
Sep 20, 2020 9:01:24 PM com.hazelcast.internal.server.tcp.TcpServerConnector
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Connecting to /127.0.0.1:5710, timeout: 10000, bind-any: true
Sep 20, 2020 9:01:24 PM com.hazelcast.internal.server.tcp.TcpServerConnection
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Initialized new cluster connection between /127.0.0.1:64665 and /127.0.0.1:5710
Sep 20, 2020 9:01:24 PM com.hazelcast.internal.cluster.ClusterService
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] 

Members {size:2, ver:4} [
	Member [127.0.0.1]:5710 - cff4c063-ebf6-458d-8bef-1f819e6947d6
	Member [127.0.0.1]:5711 - fbe0e9ae-29f9-4d04-9c7b-231fbd569316 this
]

Sep 20, 2020 9:01:24 PM com.hazelcast.core.LifecycleService
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5711 is MERGED
*** only now the cluster nodes see each other ***
```

Thanks @jerrinot for taking this up.
I am not quite sure if the steps below are a coincidence or not, because the merge timeout is also set to 5 seconds.
Secondly, this doesn't resolve the problem because in Payara I want to set that timeout to zero to reduce delays
in joining the cluster. See https://github.com/hazelcast/hazelcast/issues/17427 for more info.

The cluster join should work correctly with that value set to zero.

Also, about half the time running the second node after making your changes and bumping up the split brain timeouts to 15 seconds, this happens while starting a second node:

```
Sep 29, 2020 10:49:53 AM com.hazelcast.instance.AddressPicker
INFO: [LOCAL] [dev] [4.1-SNAPSHOT] Using public address: [127.0.0.1]:5711
Sep 29, 2020 10:49:53 AM com.hazelcast.system
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Hazelcast 4.1-SNAPSHOT (20200928 - ddebab0) starting at [127.0.0.1]:5711
Sep 29, 2020 10:49:54 AM com.hazelcast.instance.impl.Node
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Using Discovery SPI
Sep 29, 2020 10:49:54 AM com.hazelcast.cp.CPSubsystem
WARNING: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] CP Subsystem is not enabled. CP data structures will operate in UNSAFE mode! Please note that UNSAFE mode will not provide strong consistency guarantees.
Sep 29, 2020 10:49:54 AM com.hazelcast.internal.diagnostics.Diagnostics
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Diagnostics disabled. To enable add -Dhazelcast.diagnostics.enabled=true to the JVM arguments.
Sep 29, 2020 10:49:54 AM com.hazelcast.core.LifecycleService
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5711 is STARTING
Sep 29, 2020 10:49:54 AM com.hazelcast.internal.cluster.impl.DiscoveryJoiner
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5712 is added to the blacklist.
Sep 29, 2020 10:49:54 AM com.hazelcast.internal.server.tcp.TcpServerConnection
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Initialized new cluster connection between /127.0.0.1:50446 and /127.0.0.1:5710
Sep 29, 2020 10:49:54 AM com.hazelcast.internal.cluster.impl.DiscoveryJoiner
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5712 is added to the blacklist.
Sep 29, 2020 10:49:55 AM com.hazelcast.internal.cluster.impl.DiscoveryJoiner
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5712 is added to the blacklist.
Sep 29, 2020 10:49:55 AM com.hazelcast.internal.cluster.impl.DiscoveryJoiner
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5712 is added to the blacklist.
Sep 29, 2020 10:49:55 AM com.hazelcast.internal.cluster.impl.DiscoveryJoiner
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5712 is added to the blacklist.
Sep 29, 2020 10:49:59 AM com.hazelcast.instance.impl.Node
SEVERE: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Could not join cluster. Shutting down now!
Sep 29, 2020 10:49:59 AM com.hazelcast.core.LifecycleService
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5711 is SHUTTING_DOWN
Sep 29, 2020 10:49:59 AM com.hazelcast.instance.impl.Node
WARNING: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Terminating forcefully...
Sep 29, 2020 10:49:59 AM com.hazelcast.instance.impl.Node
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Shutting down connection manager...
Sep 29, 2020 10:49:59 AM com.hazelcast.internal.server.tcp.TcpServerConnection
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Connection[id=1, /127.0.0.1:50446->/127.0.0.1:5710, qualifier=null, endpoint=[127.0.0.1]:5710, alive=false, connectionType=MEMBER, planeIndex=0] closed. Reason: TcpServer is stopping
Sep 29, 2020 10:49:59 AM com.hazelcast.instance.impl.Node
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Shutting down node engine...
Sep 29, 2020 10:49:59 AM com.hazelcast.instance.impl.NodeExtension
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Destroying node NodeExtension.
Sep 29, 2020 10:49:59 AM com.hazelcast.instance.impl.Node
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Hazelcast Shutdown is completed in 17 ms.
Sep 29, 2020 10:49:59 AM com.hazelcast.core.LifecycleService
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] [127.0.0.1]:5711 is SHUTDOWN
Sep 29, 2020 10:49:59 AM com.hazelcast.instance.impl.Node
WARNING: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Config seed port is 5710 and cluster size is 1. Some of the ports seem occupied!
Sep 29, 2020 10:49:59 AM com.hazelcast.internal.util.PhoneHome
WARNING: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Could not schedule phone home task! Most probably Hazelcast failed to start.
Sep 29, 2020 10:49:59 AM com.hazelcast.instance.impl.Node
INFO: [127.0.0.1]:5711 [dev] [4.1-SNAPSHOT] Node is already shutting down... Waiting for shutdown process to complete...
[WARNING] 
java.lang.IllegalStateException: Node failed to start!
    at com.hazelcast.instance.impl.HazelcastInstanceImpl.<init> (HazelcastInstanceImpl.java:125)
    at com.hazelcast.instance.impl.HazelcastInstanceFactory.constructHazelcastInstance (HazelcastInstanceFactory.java:211)
    at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance (HazelcastInstanceFactory.java:190)
    at com.hazelcast.instance.impl.HazelcastInstanceFactory.newHazelcastInstance (HazelcastInstanceFactory.java:128)
    at com.hazelcast.core.Hazelcast.newHazelcastInstance (Hazelcast.java:57)
    at com.flowlogix.hazelcast.cachetester.CacheTester.setup (CacheTester.java:85)
    at com.flowlogix.hazelcast.cachetester.CacheTester.main (CacheTester.java:53)
    at org.codehaus.mojo.exec.ExecJavaMojo$1.run (ExecJavaMojo.java:254)
    at java.lang.Thread.run (Thread.java:748)
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  7.674 s
[INFO] Finished at: 2020-09-29T10:50:00-05:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.codehaus.mojo:exec-maven-plugin:3.0.0:java (default) on project CacheTester: An exception occured while executing the Java class. Node failed to start! -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

> hi @lprimak,
thank you for creating the reproducer, that's definitely helpful!

> This is because the CacheTester has the WAIT_SECONDS_BEFORE_JOIN property set to "0". When I comment this out and add config.setProperty(DISCOVERY_SPI_ENABLED.getName(), "true"); then I can see a newly joining member will connect to an already running cluster instead of forming its own cluster first.

--- a/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/DiscoveryJoiner.java
+++ b/hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/DiscoveryJoiner.java
@@ -28,7 +28,6 @@ import com.hazelcast.internal.util.concurrent.IdleStrategy;
 
 import java.util.ArrayList;
 import java.util.Collection;
-import java.util.Collections;
 import java.util.Set;
 
 import static com.hazelcast.spi.properties.ClusterProperty.WAIT_SECONDS_BEFORE_JOIN;
@@ -71,15 +70,16 @@ public class DiscoveryJoiner
 
     @Override
     protected Collection<Address> getPossibleAddressesForInitialJoin() {
+        Collection<Address> possibleAddresses = null;
         long deadLine = System.nanoTime() + SECONDS.toNanos(maximumWaitingTimeBeforeJoinSeconds);
         for (int i = 0; System.nanoTime() < deadLine; i++) {
-            Collection<Address> possibleAddresses = getPossibleAddresses();
+            possibleAddresses = getPossibleAddresses();
             if (!possibleAddresses.isEmpty()) {
                 return possibleAddresses;
             }
             idleStrategy.idle(i);
         }
-        return Collections.emptyList();
+        return possibleAddresses == null ? getPossibleAddresses() : possibleAddresses;
     }
 
     @Override

--- a/hazelcast/src/test/java/com/hazelcast/cluster/impl/DiscoverySPIJoinTest.java
+++ b/hazelcast/src/test/java/com/hazelcast/cluster/impl/DiscoverySPIJoinTest.java
@@ -1,0 +1,121 @@
+/*
+ * Copyright (c) 2008-2023, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hazelcast.cluster.impl;
+
+import com.hazelcast.cluster.Address;
+import com.hazelcast.cluster.ClusterState;
+import com.hazelcast.config.Config;
+import com.hazelcast.core.HazelcastInstance;
+import com.hazelcast.spi.discovery.DiscoveryNode;
+import com.hazelcast.spi.discovery.SimpleDiscoveryNode;
+import com.hazelcast.spi.discovery.integration.DiscoveryService;
+import com.hazelcast.spi.discovery.integration.DiscoveryServiceSettings;
+import com.hazelcast.test.HazelcastSerialClassRunner;
+import com.hazelcast.test.HazelcastTestSupport;
+import com.hazelcast.test.OverridePropertyRule;
+import com.hazelcast.test.annotation.QuickTest;
+import com.hazelcast.test.TestHazelcastInstanceFactory;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.experimental.categories.Category;
+import org.junit.runner.RunWith;
+import java.net.UnknownHostException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import static com.hazelcast.spi.properties.ClusterProperty.DISCOVERY_SPI_ENABLED;
+import static com.hazelcast.spi.properties.ClusterProperty.WAIT_SECONDS_BEFORE_JOIN;
+import static com.hazelcast.test.OverridePropertyRule.set;
+import static com.hazelcast.test.TestEnvironment.HAZELCAST_TEST_USE_NETWORK;
+
+/**
+ * Check non-split-brain join for Discovery SPI
+ */
+@RunWith(HazelcastSerialClassRunner.class)
+@Category(QuickTest.class)
+public class DiscoverySPIJoinTest extends HazelcastTestSupport {
+    @Rule
+    public final OverridePropertyRule overridePropertyRule = set(HAZELCAST_TEST_USE_NETWORK, "true");
+    @Rule
+    public final OverridePropertyRule overridePropertyRule2 = set(DISCOVERY_SPI_ENABLED.getName(), "true");
+    @Rule
+    public final OverridePropertyRule overridePropertyRule3 = set(WAIT_SECONDS_BEFORE_JOIN.getName(), "0");
+    private TestHazelcastInstanceFactory factory;
+
+    public static class TestDiscoveryService implements DiscoveryService  {
+        private static final List<DiscoveryNode> nodes = new ArrayList<>();
+
+        public TestDiscoveryService(DiscoveryServiceSettings settings) {
+        }
+
+        @Override
+        public void start() {
+            try {
+                nodes.add(new SimpleDiscoveryNode(new Address("127.0.0.1", 5701)));
+                nodes.add(new SimpleDiscoveryNode(new Address("127.0.0.1", 5702)));
+                nodes.add(new SimpleDiscoveryNode(new Address("127.0.0.1", 5703)));
+            } catch (UnknownHostException ex) {
+                throw new RuntimeException(ex);
+            }
+        }
+
+        @Override
+        public Iterable<DiscoveryNode> discoverNodes() {
+            return Collections.unmodifiableList(nodes);
+        }
+
+        @Override
+        public void destroy() {
+            nodes.clear();
+        }
+
+        @Override
+        public Map<String, String> discoverLocalMetadata() {
+            return Collections.unmodifiableMap(new HashMap<>());
+        }
+    }
+
+    @Before
+    public void beforeRun() {
+        factory = createHazelcastInstanceFactory(2);
+    }
+
+    @After
+    public void afterRun() {
+        factory.shutdownAll();
+    }
+
+    @Override
+    protected Config getConfig() {
+        Config config = new Config();
+        config.getNetworkConfig().getJoin().getMulticastConfig().setEnabled(false);
+        config.getNetworkConfig().getJoin()
+            .getDiscoveryConfig().setDiscoveryServiceProvider(TestDiscoveryService::new);
+        return config;
+    }
+
+    @Test
+    public void makeSureNoSplitBrain() {
+        HazelcastInstance hz1 = factory.newHazelcastInstance(getConfig());
+        HazelcastInstance hz2 = factory.newHazelcastInstance(getConfig());
+        assertClusterStateEventually(ClusterState.ACTIVE, hz1, hz2);
+        assertClusterSize(2, hz1, hz2);
+    }
+}

